[Shipping labels] Migrate shipping labels eligibility endpoint

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-655

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the eligibility endpoint for shipping labels and uses it on the order detail screen when the feature flag is enabled.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Go to the Orders screen.
2. Open an order that is eligible for shipping labels.
3. Repeat with an order that is not eligible for shipping labels. 

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
I also tested this for disabled flag case.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add EligibilityResponse DTO

irfano committed:
Add `fetchShippingEligibility` endpoint to `WooShippingLabelRestClient`

irfano committed:
Add repository and data store functions for shipping label eligibility

irfano committed:
Update shipping label creation check in OrderDetailViewModel

irfano committed:
Update OrderDetailViewModelTest

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/261">22.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingLabelRestClient.kt:
```diff
@@ -18,6 +18,16 @@ import javax.inject.Inject
+            site = site,
+            path = url,
+            clazz = EligibilityResponse::class.java,
+        ).toWooPayload()
```
I didn't add `can_create_customs_form:true`, `can_create_package:true` and `can_create_payment_method:true` parameters as their default value is already `true`.

hichamboushaba replied:
I think we should still pass them, just in case the plugin changes its logic or something in the future, WDYT?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/datasource/WooShippingEligibilityDataStore.kt:
```diff
@@ -0,0 +1,34 @@
+import javax.inject.Inject
+
+class WooShippingEligibilityDataStore @Inject constructor(
+    @DataStoreQualifier(DataStoreType.SHIPPING_LABELS_DATA) private val dataStore: DataStore<Preferences>,
```
I have a remark here (not blocking, and doesn't require changes now), I was planning to have this discussion in M5, but given we are adding a new data to the DataStore, I'm sharing it now, and we can discuss later in M5.

I think using a DB for those data will be better and safer, as it's best suited for storing big numbers of rows.

When I started working on the project, I thought we were using the `DataStore` only for store specific data (`accountSettings` and "origin addresses`), and I wasn't familiar with the `config` endpoint, but it turns out the `config` endpoint returns a big response, and it's different for each order, so for stores with hundreds or thousands of orders, we'll end up saving all of the data in a single file, which could be problematic.

I suggest we try to migrate the `config` and this `eligibility` to the DB in M5 to avoid any potential issues, WDYT?

irfano replied:
I agree. We can do this in Milestone 5.

hichamboushaba replied:
Logged WOOMOB-726 for this.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:
```diff
@@ -948,6 +950,15 @@ class OrderDetailViewModel @Inject constructor(
 
+    private suspend fun isOrderEligibleForSLCreation(orderEligibleForInPersonPayments: Boolean) =
+        if (FeatureFlag.REVAMP_WOO_SHIPPING.isEnabled()) {
+            shippingLabelOnboardingRepository.shippingPluginSupport.isWooShippingSupported() &&
```
This breaks the old flow when the feature flag is enabled, because we then check only the new plugin eligibility and not the old one. We can fix this by using `isRevampWooShippingEnabled` in the `if` condition.

(You can test this by disabling the new plugin, the "create shipping label" button will not be shown)

irfano replied:
Great catch! We should indeed show the old flow if the new plugin is not installed but the old one is. My implementation was incorrect, but I’ve now fixed the logic: 4543a2bf24726bd6555975de001d635f7c3c862a

hichamboushaba reviewed (commented):
Nice work, there is one blocking [issue](https://github.com/woocommerce/woocommerce-android/pull/14250#discussion_r2172541998) though, please check the comments 🙏

irfano committed:
Pass new parameters for WooShipping eligibility check

This commit modifies the `WooShippingLabelRestClient` to pass three new boolean parameters (`can_create_customs_form`, `can_create_package`, `can_create_payment_method`) to the eligibility check endpoint. These parameters are all set to `true`.

irfano committed:
Fix shipping label eligibility logic

hichamboushaba reviewed (approved):
Thanks for fixing the issue @irfano, nice work :shipit:
